### PR TITLE
Securely encrypts reset token before storing in the database

### PR DESCRIPTION
**Pull request by codingfriend1**
### Problem

Any hacker with read access to the database can change passwords easily just by sending a request and reading the raw reset token for the user. @micaksica2 wrote an article describing this problem and recommending to securely encrypt reset tokens.

https://hackernoon.com/your-node-js-authentication-tutorial-is-wrong-f1a3bf831a46

### Summary

Now the short and long reset tokens are first sent in the notifier and then encrypted (using the same password encryption method) before being stored in the database. When the user makes the reset password request it will bcrypt compare the reset token and the encrypted one stored in the database using `passwordCompare`.

NOTE: Before the reset token was used to query for users by itself, since it's now encrypted this method will no longer work. When creating a token we will concatenate the user id with the token (separated by 3 `___`) and extract the id out to query the user (not sure if this is a security issue).

Issue #64 

https://github.com/feathersjs/feathers-authentication-management/issues/64

### Followup

This will involve a correction in the docs that every token, short or long will be stored in the database as strings with 60 characters length. However, the longer or shorter tokens may be sent to the notifier before encrypted.

_WARNING: **This will break existing reset tokens forcing the users to make the reset password request again.** It will also require us to **throttle reset requests** or limit the number of reset password requests that can be made within a time frame so a hacker couldn't slow down our server performance by running thousands of reset password requests._

### CODE REVIEW

**Comment by eddyystop** [comment]
I will review this as soon as I have some time.

File: test/helpers/feathersStubs.js [hunk 1]
```diff
@@ -62,6 +62,17 @@ module.exports.users = function users (app, usersDb, nonPaginated, idProp = '_id
...
       });
     },
+    get (id) { // always use as a Promise
+      const index = usersDb.findIndex(user1 => user1[idProp] === id);
+
+      if (index === -1) {
+        return Promise.reject(new Error(`users.get ${idProp}=${id} not found.`));
+      }
+
+      debug('/users get: %d %o', usersDb[index], id);
```
Comment (micaksica2) [inline test/helpers/feathersStubs.js]: Remove your debug statement maybe?

**Comment by micaksica2** [comment]
Hm. I haven't looked at this yet but I thought I'd chime in.

Quick thoughts on the id___token construct: without TLS you have an information disclosure on the wire that the token belongs to UID `id` but I don't think that's a big deal. I can't think of any way with the way Feathers currently works in such that this will be an issue, unless you can chain it with something else. Using `id__token` now queries a user by id, then checks the hash of the token. `id` lookups aren't injectable, and I can't think of a situation where this construct is worse off than just `token`. Actually,  in the case of predictable tokens the `id__token` construct will now require an attacker to not just guess a token, but guess which user the token belongs to to gain access to an account.

Using bcrypt on tokens will slow down brute force but adds a DoS vector as you stated unless the endpoints for reset are rate limited. I'll think about this a bit and see what I can come up with that may be a harmful side effect here.

**Comment by codingfriend1** [comment]
@micaksica2 In the case this isn't over TLS could having the id provide enough information for the attacker to find which email is associated with the password being changed? Is that information he would already have though? Especially if he had read access to the database?

**Comment by eddyystop** [comment]
@codingfriend1 and @micaksica2 , are you both happy with this PR? If so, I'll publish it later today.

**Comment by codingfriend1** [comment]
I am. I would like somebody besides me to run some real world tests on it though before we commit. I ran some myself but it's better to have a crowd.

**Comment by eddyystop** [comment]
Sorry for the long delay in publishing this as v1.0.0.

**Pull request merged**